Update tidb-distributed-execution-framework.md (#15606)
pingcap/docs: tidb-distributed-execution-framework.md

> ---

Author: Added

> title: TiDB Backend Task Distributed Execution Framework (DXF)

Reviewer: ```suggestion
title: TiDB Backend Task Distributed eXecution Framework (DXF)
```

> summary: Learn the use cases, limitations, usage, and implementation principles of the TiDB backend task distributed execution  (DXF).

Reviewer: ```suggestion
summary: Learn the use cases, limitations, usage, and implementation principles of the TiDB backend task distributed execution framework (DXF).
```

</CustomContent>

TiDB adopts a computing-storage separation architecture with excellent scalability and elasticity. Starting from v7.1.0, TiDB introduces a backend task DXF (Distributed eXecution Framework) to further leverage the resource advantages of the distributed architecture. The goal of this framework is to implement unified scheduling and distributed execution of all backend tasks, and to provide unified resource management capabilities for both overall and individual backend tasks, which better meets users' expectations for resource usage.

> This document describes the use cases, limitations, usage, and implementation principles of the DXF.

Reviewer 2: ```suggestion
This document describes the use cases, usage, and implementation principles of the DXF.
```

</CustomContent>

<CustomContent platform="tidb-cloud">

In a database management system, in addition to the core transactional processing (TP) and analytical processing (AP) workloads, there are other important tasks, such as DDL operations, TTL, Analyze, and Backup/Restore. These tasks need to process a large amount of data in database objects (tables), so they typically have the following characteristics:

</CustomContent>

- Need to process all data in a schema or a database object (table).
- Might need to be executed periodically, but at a low frequency.
- If the resources are not properly controlled, they are prone to affect TP and AP tasks, lowering the database service quality.

Enabling the DXF can solve the above problems and has the following three advantages:

- The framework provides unified capabilities for high scalability, high availability, and high performance.

> - The framework supports distributed execution of tasks, which can flexibly schedule the available computing resources of the entire TiDB cluster, thereby better utilizing the computing resources in a TiDB cluster.
> - The framework provides unified resource usage and management capabilities for both overall and individual tasks.

Reviewer: ```suggestion
- The DXF supports distributed execution of tasks, which can flexibly schedule the available computing resources of the entire TiDB cluster, thereby better utilizing the computing resources in a TiDB cluster.
- The DXF provides unified resource usage and management capabilities for both overall and individual tasks.
```

<CustomContent platform="tidb-cloud">

> Currently, for TiDB Self-Hosted, the DXF supports the distributed execution of the `ADD INDEX`. `ADD INDEX` is a DDL statement used to create indexes. For example:

Reviewer: ```suggestion
Currently, the DXF supports the distributed execution of the `ADD INDEX`. `ADD INDEX` is a DDL statement used to create indexes. For example:
```

```sql
ALTER TABLE t1 ADD INDEX idx1(c1);
CREATE INDEX idx1 ON table t1(c1);
```

</CustomContent>

<CustomContent platform="tidb">

> Currently, for TiDB Self-Hosted, the DXF supports the distributed execution of the `ADD INDEX` and `IMPORT INTO` statements. For TiDB Cloud, the `IMPORT INTO` statement is not applicable until v7.2.

Reviewer: ```suggestion
Currently, for TiDB Self-Hosted, the DXF supports the distributed execution of the `ADD INDEX` and `IMPORT INTO` statements.
```

Reviewer: However, it is duplicated with line 46

- Might need to be executed periodically, but at a low frequency.
- If the resources are not properly controlled, they are prone to affect TP and AP tasks, lowering the database service quality.

Enabling the DXF can solve the above problems and has the following three advantages:

- The framework provides unified capabilities for high scalability, high availability, and high performance.

> - The framework supports distributed execution of backend tasks, which can flexibly schedule the available computing resources of the entire TiDB cluster, thereby better utilizing the computing resources in a TiDB cluster.

Reviewer 3: ```suggestion
- The framework supports distributed execution of tasks, which can flexibly schedule the available computing resources of the entire TiDB cluster, thereby better utilizing the computing resources in a TiDB cluster.
```

> - The framework provides unified resource usage and management capabilities for both overall and individual backend tasks.

Reviewer 3: ```suggestion
- The framework provides unified resource usage and management capabilities for both overall and individual tasks.
```

> Currently, for TiDB Self-Hosted, the DXF supports the distributed execution of the `ADD INDEX` and `IMPORT INTO` statements. For TiDB Cloud, the `IMPORT INTO` statement is not applicable until v7.2.

Reviewer 3: The `IMPORT INTO` statement is not supported on TiDB Cloud.

Author: Updated

- `IMPORT INTO` is used to import data in formats such as `CSV`, `SQL`, and `PARQUET` into an empty table. For more information, see [`IMPORT INTO`](https://docs.pingcap.com/tidb/v7.2/sql-statement-import-into).

## Limitation

> The DXF can only schedule one `ADD INDEX` task at a time. If a second `ADD INDEX` task is submitted before the first `ADD INDEX` task is completed, the second `ADD INDEX` task is executed through a transaction instead of the DXF.

Author: ```suggestion
The DXF can only schedule the distributed execution of one `ADD INDEX` task at a time. If a new `ADD INDEX` task is submitted before the current `ADD INDEX` distributed task has finished, the new task is executed through a transaction.
```

## Usage

1. To enable the DXF, set the value of [`tidb_enable_dist_task`](/system-variables.md#tidb_enable_dist_task-new-in-v710) to `ON`:

>     ```sql

Reviewer 3: L114 和 L120 中的 background tasks 需要修改为 tasks

    </CustomContent>

> 2. It is recommended that you use their default values for the following system variables that might affect the distributed execution of DDL tasks:

Reviewer 3: ```suggestion
2. For the following system variables that might affect the distributed execution of DDL tasks, it is recommended that you use their default values:
```

    </CustomContent>

> 2. For the following system variables that might affect the distributed execution of DDL tasks, it is recommended that you use their default values:

Reviewer: ```suggestion
2. In general, for the following system variables that might affect the distributed execution of DDL tasks, it is recommended that you use their default values:
```

    * [`tidb_ddl_reorg_worker_cnt`](/system-variables.md#tidb_ddl_reorg_worker_cnt): use the default value `4`. The recommended maximum value is `16`.
    * [`tidb_ddl_reorg_priority`](/system-variables.md#tidb_ddl_reorg_priority)
    * [`tidb_ddl_error_count_limit`](/system-variables.md#tidb_ddl_error_count_limit)
    * [`tidb_ddl_reorg_batch_size`](/system-variables.md#tidb_ddl_reorg_batch_size): use the default value. The recommended maximum value is `1024`.

> 3. Starting from v7.4.0, you can adjust the number of TiDB nodes that perform background tasks according to actual needs. After deploying a TiDB cluster, you can set the instance-level system variable [`tidb_service_scope`](/system-variables.md#tidb_service_scope-new-in-v740) for each TiDB node in the cluster. When `tidb_service_scope` of a TiDB node is set to `background`, the TiDB node can execute background tasks. When `tidb_service_scope` of a TiDB node is set to the default value "", the TiDB node cannot execute background tasks. If `tidb_service_scope` is not set for any TiDB node in a cluster, the DXF schedules all TiDB nodes to execute background tasks by default.

Reviewer: ```suggestion
3. Starting from v7.4.0, you can adjust the number of TiDB nodes that perform the DXF tasks according to actual needs. After deploying a TiDB cluster, you can set the instance-level system variable [`tidb_service_scope`](/system-variables.md#tidb_service_scope-new-in-v740) for each TiDB node in the cluster. When `tidb_service_scope` of a TiDB node is set to `background`, the TiDB node can execute the DXF tasks. When `tidb_service_scope` of a TiDB node is set to the default value "", the TiDB node cannot execute DXF tasks. If `tidb_service_scope` is not set for any TiDB node in a cluster, the DXF schedules all TiDB nodes to execute background tasks by default.
```
